In [66]:
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt

In [67]:
df=pd.read_csv('/California.zip')

In [68]:
df.info()
df.shape
df.describe()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 20640 entries, 0 to 20639
Data columns (total 10 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   longitude           20640 non-null  float64
 1   latitude            20640 non-null  float64
 2   housing_median_age  20640 non-null  float64
 3   total_rooms         20640 non-null  float64
 4   total_bedrooms      20433 non-null  float64
 5   population          20640 non-null  float64
 6   households          20640 non-null  float64
 7   median_income       20640 non-null  float64
 8   median_house_value  20640 non-null  float64
 9   ocean_proximity     20640 non-null  object 
dtypes: float64(9), object(1)
memory usage: 1.6+ MB


,longitude,latitude,housing_median_age,total_rooms,total_bedrooms,population,households,median_income,median_house_value
count,20640.000000,20640.000000,20640.000000,20640.000000,20433.000000,20640.000000,20640.000000,20640.000000,20640.000000
mean,-119.569704,35.631861,28.639486,2635.763081,537.870553,1425.476744,499.539680,3.870671,206855.816909
std,2.003532,2.135952,12.585558,2181.615252,421.385070,1132.462122,382.329753,1.899822,115395.615874
min,-124.350000,32.540000,1.000000,2.000000,1.000000,3.000000,1.000000,0.499900,14999.000000
25%,-121.800000,33.930000,18.000000,1447.750000,296.000000,787.000000,280.000000,2.563400,119600.000000
50%,-118.490000,34.260000,29.000000,2127.000000,435.000000,1166.000000,409.000000,3.534800,179700.000000
75%,-118.010000,37.710000,37.000000,3148.000000,647.000000,1725.000000,605.000000,4.743250,264725.000000
max,-114.310000,41.950000,52.000000,39320.000000,6445.000000,35682.000000,6082.000000,15.000100,500001.000000


In [69]:
df.duplicated().sum()

np.int64(0)

In [70]:
df.head()

,longitude,latitude,housing_median_age,total_rooms,total_bedrooms,population,households,median_income,median_house_value,ocean_proximity
0,-122.23,37.88,41.0,880.0,129.0,322.0,126.0,8.3252,452600.0,NEAR BAY
1,-122.22,37.86,21.0,7099.0,1106.0,2401.0,1138.0,8.3014,358500.0,NEAR BAY
2,-122.24,37.85,52.0,1467.0,190.0,496.0,177.0,7.2574,352100.0,NEAR BAY
3,-122.25,37.85,52.0,1274.0,235.0,558.0,219.0,5.6431,341300.0,NEAR BAY
4,-122.25,37.85,52.0,1627.0,280.0,565.0,259.0,3.8462,342200.0,NEAR BAY


In [71]:
df.isnull().sum()

,0
longitude,0
latitude,0
housing_median_age,0
total_rooms,0
total_bedrooms,207
population,0
households,0
median_income,0
median_house_value,0
ocean_proximity,0


In [72]:
df['total_bedrooms']=df['total_bedrooms'].fillna(df['total_bedrooms'].mean())

In [73]:
df = pd.get_dummies(df,columns=['ocean_proximity'],drop_first=True)

In [74]:
df.head()

,longitude,latitude,housing_median_age,total_rooms,total_bedrooms,population,households,median_income,median_house_value,ocean_proximity_INLAND,ocean_proximity_ISLAND,ocean_proximity_NEAR BAY,ocean_proximity_NEAR OCEAN
0,-122.23,37.88,41.0,880.0,129.0,322.0,126.0,8.3252,452600.0,False,False,True,False
1,-122.22,37.86,21.0,7099.0,1106.0,2401.0,1138.0,8.3014,358500.0,False,False,True,False
2,-122.24,37.85,52.0,1467.0,190.0,496.0,177.0,7.2574,352100.0,False,False,True,False
3,-122.25,37.85,52.0,1274.0,235.0,558.0,219.0,5.6431,341300.0,False,False,True,False
4,-122.25,37.85,52.0,1627.0,280.0,565.0,259.0,3.8462,342200.0,False,False,True,False


In [75]:
x=df.drop('median_house_value',axis=1)
y=df['median_house_value']

In [76]:
from sklearn.model_selection import train_test_split
x_train,x_test,y_train,y_test=train_test_split(x,y,test_size=0.2,random_state=42)

In [77]:
from sklearn.preprocessing import StandardScaler
SS=StandardScaler()

In [78]:
x_train=SS.fit_transform(x_train)
x_test=SS.transform(x_test)

In [79]:
# Linear Regression
from sklearn.linear_model import LinearRegression
LR=LinearRegression()
LR.fit(x_train,y_train)

LinearRegression()

In [80]:
y_pred_lr=LR.predict(x_test)

In [81]:
from sklearn.metrics import mean_squared_error,root_mean_squared_error,mean_absolute_error,r2_score

In [82]:
LR_MAE = mean_absolute_error(y_test, y_pred_lr)
LR_MSE = mean_squared_error(y_test, y_pred_lr)
LR_RMSE = root_mean_squared_error(y_test, y_pred_lr)
LR_R2 = r2_score(y_test, y_pred_lr)

In [83]:
print("Linear Regression Values")
print("MAE:", LR_MAE)
print("MSE:", LR_MSE)
print("RMSE:", LR_RMSE)
print("R2:", LR_R2)

Linear Regression Values
MAE: 50701.77903132993
MSE: 4904399775.949275
RMSE: 70031.41991955665
R2: 0.6257351821159706


In [84]:
LR.coef_

array([-53826.64801649, -54415.6961445 ,  13889.86618856, -13094.25116219,
        43068.18184187, -43403.43242732,  18382.19632373,  75167.77476625,
       -18506.09522509,   2118.43757154,  -1621.70501786,   1139.90559187])

In [85]:
LR.intercept_

np.float64(207194.69373788772)

In [86]:
# 1. Lasso Regression
from sklearn.linear_model import Lasso

L1=Lasso(alpha=1)
L1.fit(x_train,y_train)

y_pred_lasso=L1.predict(x_test)

Lasso_MAE = mean_absolute_error(y_test, y_pred_lasso)
Lasso_MSE = mean_squared_error(y_test, y_pred_lasso)
Lasso_RMSE = root_mean_squared_error(y_test, y_pred_lasso)
Lasso_R2 = r2_score(y_test, y_pred_lasso)

print("Lasso Regression Values")
print("MAE:", Lasso_MAE)
print("MSE:", Lasso_MSE)
print("RMSE:", Lasso_RMSE)
print("R2:", Lasso_R2)

Lasso Regression Values
MAE: 50701.195542048394
MSE: 4904310967.020979
RMSE: 70030.7858518022
R2: 0.6257419593076543


In [87]:
L1.coef_

array([-53797.75028931, -54386.67369765,  13889.19876917, -13073.22587596,
        43045.41252989, -43396.11340072,  18377.55734347,  75162.82671862,
       -18514.69246299,   2117.64610678,  -1618.42318951,   1140.96135222])

In [88]:
L1.intercept_

np.float64(207194.69373788772)

In [89]:
# 2. Ridge Regression
from sklearn.linear_model import Ridge
L2=Ridge(alpha=1)
L2.fit(x_train,y_train)
y_pred_ridge=L2.predict(x_test)
Ridge_MAE = mean_absolute_error(y_test, y_pred_ridge)
Ridge_MSE = mean_squared_error(y_test, y_pred_ridge)
Ridge_RMSE = root_mean_squared_error(y_test, y_pred_ridge)
Ridge_R2 = r2_score(y_test, y_pred_ridge)
print("Ridge Regression")
print("MAE:", Ridge_MAE)
print("MSE:", Ridge_MSE)
print("RMSE:", Ridge_RMSE)
print("R2:", Ridge_R2)

Ridge Regression
MAE: 50699.22919657207
MSE: 4903987826.86813
RMSE: 70028.47868451898
R2: 0.625766618796273


In [90]:
L2.coef_

array([-53713.65159257, -54298.40041519,  13890.4723292 , -13058.64374116,
        42978.32840348, -43389.96877929,  18425.41255042,  75157.63575176,
       -18538.72553023,   2119.24916563,  -1616.22226055,   1146.56402572])

In [91]:
L2.intercept_

np.float64(207194.69373788772)

1. To choose a best fit model:MAE,MSE,RMSE should be low and R2 score should be higher.
2. Among the three models, Ridge Regression performed best on the test data. It produced the lowest MAE (50,699.23), MSE (4.904 billion), and RMSE (70,028.48), while achieving the highest R² score of 0.6258. Lasso and Linear Regression produced almost identical results, but Ridge performed slightly better across all evaluation metrics.
3. Ridge regularization shrinks the model coefficients, which can help reduce overfitting. Therefore, based on the test-set results, Ridge Regression generalizes best among the three models.